### PR TITLE
Extend connectivity-check for HTTP policy validation via CUE

### PR DESCRIPTION
**Pull request by joestringer**
This PR introduces a [CUE](https://cuelang.org/)-based framework for defining connectivity checks to consolidate the definitions of the various connectivity checks and support simpler, more concise declarations of new connectivity checks. 

Review commit-by-commit.

### Why CUE?

CUE is a data constraint language built defined as a superset of JSON which aims to "simplify tasks involving defining and using data".  See the second commit for a more detailed discussion, talking points below as a summary:

    In context of the connectivity check YAMLs, CUE is useful to allow us to
    "evaporate" the boilerplate necessary to define Kubernetes YAMLs for
    Deployments, Services and CiliumNetworkPolicies and allow developers to
    specify various permutations for connectivity checks concisely.

    Why should we use it?
    * It's more concise. [...]
    * Consolidate constant declaration. [...]
    * More structured data declarations for k8s resources of checks [...]. Can build powerful tooilng around this.

    What are the dangers (+mitigations)?
    * It's relatively imature, v0.2.2. This PR pins CUE usage to this version. [...]
    * Another language in the tree. [...] This however aligns with Go and k8s community exploration of the language.
    * It's easy to hide all complexity, templating away all parameters in a way that makes the declarations implicit. It'll be up to us to figure out how explicitly to declare our check definitions. [...]
    * Additive declarations. Evaluation of language statements are not ordered by line ordering. Can be counter-intuitive. [...]
 
Based on the above and the experience of converting these checks over to CUE, I believe this is a useful and powerful tool for declaring YAML variations, perfect for a use case like connectivity-checks. We may in future even be able to explore using CUE for defining manifests in the CI.

### How do we know these changes are correct?

This PR was validated against the existing "hostport" connectivity check YAML from the tree using the `make inspect` target introduced in this PR. I've attached the diff for existing checks below for reference; the commit message for the relevant changes also highlights the specifics of what has generally changed:

     * All containers are converted to use readinessProbe and not
       livenessProbe.
     * All readiness probes now specify --connect-timeout of 5s.
     * Readiness probes access `/public` or `/private` per the underlying
       container HTTP server paths rather than just accessing `/`.
     * DNS allow policies are converted to consistently allow both TCP and
       UDP-based DNS.
     * Container names are derived from pod names.
     * The new YAMLs declare additional labels for all resourcess, such as
       'component' and 'topology'.

[connectivity-check-hostport.yaml.diff.txt](https://github.com/cilium/cilium/files/4973710/connectivity-check-hostport.yaml.diff.txt)

Tested on a 2-node GKE cluster per the Cilium getting-started guide.

### Extending the connectivity checks for HTTP policy validation

See the commit log for details:

    These new checks configure various L7 proxy paths to validate
    connectivity via L7 proxies, in the following dimensions:

    - Apply policy on egress; ingress; or both (proxy-to-proxy)
    - Intra-node / Multi-node
    - Allow / Deny

    Note that proxy-to-proxy always configures egress allow policy to ensure
    that the traffic goes via the proxy and in the drop case the requests
    are only rejected at the destination. This is because applying egress
    deny at the source would prevent proxy-to-proxy connectivity, meaning
    the test would be equivalent to the egress-only reject policy case. This
    way, we ensure that the path via the egress proxy to the destination is
    tested in the reject case.

    These are implemented partially through a new 'echo-c' pod which always
    has ingress policy applied to allow GET requests to '/public'. Depending
    on whether ingress policy is needed to check the particular permutation
    the new checks may connect to 'echo-a' or 'echo-c'.

    These are implemented by adding pods for each permutation of policy
    apply point and topology; then by adding allow / deny containers within
    that pod to test the allow/deny cases.

    The 'connectivity-check-proxy.yaml' includes all of the above.

    Finally, the omissions: This commit does not attempt to address
    variations in datapath configuration. This includes IPv4 vs. IPv6;
    tunnel/direct-routing; endpoint config; kube proxy/free; encryption.
    These are left up to the cluster operator configuring Cilium in specific
    modes and subsequently deploying these YAMLs.

### Tooling

One interesting aspect around CUE is the ability to write tailored CLIs against the data to filter the data you want and output in the format you're after. See the commit message for more details of the new tooling:

      $ cue cmd help
      List connectivity-check resources specified in this directory

      Usage:
        cue [-t component=<component>] [-t name=<name>] [-t topology=<topology>] <command>

      Available Commands:
        dump   Generate connectivity-check YAMLs from the cuelang scripts
        ls     List connectivity-check resources specified in this directory

    List available connectivity-check components:

      $ cue cmd ls
      KIND                  COMPONENT        TOPOLOGY     NAME
      Service               network-check    any          echo-a
      Service               services-check   any          echo-b
      Service               services-check   any          echo-b-headless
      Service               services-check   any          echo-b-host-headless
      Deployment            network-check    any          echo-a
      Deployment            services-check   any          echo-b
      Deployment            services-check   any          echo-b-host
      Deployment            network-check    any          pod-to-a
      Deployment            network-check    any          pod-to-external-1111
      Deployment            policy-check     any          pod-to-a-allowed-cnp
      Deployment            policy-check     any          pod-to-a-denied-cnp
      Deployment            policy-check     any          pod-to-external-fqdn-allow-google-cnp
      Deployment            services-check   multi-node   pod-to-b-multi-node-clusterip
      Deployment            services-check   multi-node   pod-to-b-multi-node-headless
      Deployment            services-check   intra-node   pod-to-b-intra-node-clusterip
      Deployment            services-check   intra-node   pod-to-b-intra-node-headless
      Deployment            services-check   multi-node   host-to-b-multi-node-clusterip
      Deployment            services-check   multi-node   host-to-b-multi-node-headless
      CiliumNetworkPolicy   policy-check     any          pod-to-a-allowed-cnp
      CiliumNetworkPolicy   policy-check     any          pod-to-a-denied-cnp
      CiliumNetworkPolicy   policy-check     any          pod-to-external-fqdn-allow-google-cnp

    These can be filtered by component, topology or name. For example:

      $ cue cmd -t component=network ls
      KIND         COMPONENT       TOPOLOGY    NAME
      Service      network-check   any         echo-a
      Deployment   network-check   any         echo-a
      Deployment   network-check   any         pod-to-a
      Deployment   network-check   any         pod-to-external-1111

    Finally, to gather the (filtered) YAMLs for the specified resources:

      $ cue cmd dump | head -n 20
      metadata:
        name: echo-a
        labels:
          name: echo-a
          topology: any
          component: network-check
      spec:
        ports:
          - port: 80
        selector:
          name: echo-a
        type: ClusterIP
      apiVersion: v1
      kind: Service
      ---
      ...

#### Q: Do I need to install CUE?

No, the Makefile introduces usage of the cue docker image to run the YAML generation commands. For more advanced usage like the tooling described above however, you would need to either install CUE or run it via the docker container in a similar way to how the makefile does.

### Outstanding questions

* For now I have introduced the new checks (proxy-based and hostport+proxy which are quarantined) into separate YAMLs to allow them to be tested independently. They are not part of the simple YAML. This allows the regular connectivity-check YAML to remain small and contained to allow it to reliably pass in smaller environments. We should consider how to expand the testing of these YAMLs in CI.

Related: #8491

### CODE REVIEW

**Comment by joestringer** [comment]
test-focus K8sConformance.*connectivity

**Comment by joestringer** [comment]
> This is because applying egress deny at the source would prevent proxy-to-proxy connectivity.

Well, it's also because then it would be exactly the same as the egress-only reject policy. I should update the commit message to reflect this.

**Comment by joestringer** [comment]
The two tests that use this connectivity-check passed:
* https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated-Focus/348/testReport/Suite-k8s-1/18/K8sConformance_Portmap_Chaining_Check_connectivity_check_compliance_with_portmap_chaining/
* https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated-Focus/348/testReport/Suite-k8s-1/18/K8sConformance_Portmap_Chaining_Check_one_node_connectivity_check_compliance_with_portmap_chaining/

So from a CI perspective this should be good to go. Only question might be whether we want to add a dedicated CI test for this so we can more easily focus on just the proxy failures if a regression is introduced.

**Review by jrajahalme — APPROVED** [review]
The general approach looks good, did not really review the yamls in detail. There are some additional dimensions that have been problematic in the past:

- Service access via hostport/not
- Hostport backend POD in the same v.s. other node as the one that was addressed.

Then there are bunch of datapath config dimensions (tunnel/no tunnel, endpoint routes, kube proxy/free, encryption/no, etc. ), but these are out of scope for the connectivity checks that apply to any given config.

File: examples/kubernetes/connectivity-check/pod-to-c-intra-node-proxy-ingress-reject.yaml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+        name: pod-to-c-intra-node-proxy-ingress-reject
+    spec:
+      containers:
+      - name: pod-to-c-intra-node-proxy-ingress-reject-container
+        image: docker.io/byrnedo/alpine-curl:0.1.8
+        command: ["/bin/ash", "-c", "sleep 1000000000"]
+        imagePullPolicy: IfNotPresent
+        livenessProbe:
+          exec:
+            command: ["ash", "-c", "! curl -sS --fail --connect-timeout 5 -o /dev/null echo-c/invalid-path"]
```
Comment (kkourt) [inline examples/kubernetes/connectivity-check/pod-to-c-intra-node-proxy-ingress-reject.yaml]: I wonder if it would be better to check both allow and reject in a single yaml file since it is the same policy. I.e., do something like: `curl ... echo-c/ && ! curl ... echo-c/invalid-path`

Nit: `invalid-path` was a bit confusing for me because it was not clear that it would be served by the server. Maybe `/forbidden-path` is better?

**Review by kkourt — APPROVED** [review]

File: examples/kubernetes/connectivity-check/connectivity-check-single-node.yaml [hunk 2]
```diff
@@ -268,45 +268,6 @@ spec:
 ---
 apiVersion: apps/v1
 kind: Deployment
-metadata:
```
Comment (aanm) [inline examples/kubernetes/connectivity-check/connectivity-check-single-node.yaml:271]: Is this change intentional in this commit?

**Review by aanm — COMMENTED** [review]
Minor nit

File: examples/kubernetes/connectivity-check/connectivity-check-single-node.yaml (hunk 2 above)
Comment (joestringer) [reply examples/kubernetes/connectivity-check/connectivity-check-single-node.yaml:271]: Kinda, yeah. I mean, I made modifications to the Makefile, then I re-ran `make -C examples/kubernetes/connectivity-check/ clean all` and it removed this section.

Seems to me like from the Makefile, the `single-node` variation in the connectivity check is not supposed to include any hostport checks, and yet the file as checked in has them. I'm guessing we meant to re-generate these YAMLs after an earlier commit but we never did.

I could move this blob out to a separate commit if you think that's meaningful. I didn't want to try to run the above make command before adding the `sort` because I wasn't sure that the file listing order is consistent across different machines.

File: examples/kubernetes/connectivity-check/pod-to-c-intra-node-proxy-ingress-reject.yaml (hunk 1 above)
Comment (joestringer) [reply examples/kubernetes/connectivity-check/pod-to-c-intra-node-proxy-ingress-reject.yaml]: > I wonder if it would be better to check both allow and reject in a single yaml file since it is the same policy. I.e., do something like: curl ... echo-c/ && ! curl ... echo-c/invalid-path

Single YAML file or single deployment? The idea behind having them report separately in different pods is that we can more easily pinpoint what type of issue it is, rather than "policy doesn't work between these peers" it's "policy *allow* doesn't work between these peers" or "policy *reject* doesn't work between these peers".  That said if we're concerned about deploying too many pods then this could be an easy way to halve the number of these policy-related pods. The trick would be in how we ensure that the `livenessProbe` could easily reflect which of the two errors have occurred.

> Nit: invalid-path was a bit confusing for me because it was not clear that it would be served by the server. Maybe /forbidden-path is better?

Sure, I'll fix this up.

Comment (joestringer) [reply examples/kubernetes/connectivity-check/pod-to-c-intra-node-proxy-ingress-reject.yaml]: > The idea behind having them report separately in different pods is that we can more easily pinpoint what type of issue it is

Another possibility brought up by @tgraf on Slack is to have multiple containers in the same pod.

File: examples/kubernetes/connectivity-check/echo-c.yaml [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  type: ClusterIP
+  ports:
+  - port: 80
+  selector:
+    name: echo-c
+---
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: echo-c
```
Comment (sayboras) [inline examples/kubernetes/connectivity-check/echo-c.yaml]: nit: conformance test is passed right now, but we might need to add steps to dump details and logs in github action. 
This https://github.com/cilium/cilium/blob/master/.github/workflows/smoke-test.yaml#L153 is my naive attempt, and little verbose.

File: examples/kubernetes/connectivity-check/connectivity-check-single-node.yaml (hunk 2 above)
Comment (joestringer) [reply examples/kubernetes/connectivity-check/connectivity-check-single-node.yaml:271]: FYI I removed this commit with the new approach, and validated the diff against a deployment of the existing connectivity check. I can provide that diff when I update the PR description.

File: examples/kubernetes/connectivity-check/pod-to-c-intra-node-proxy-ingress-reject.yaml (hunk 1 above)
Comment (joestringer) [reply examples/kubernetes/connectivity-check/pod-to-c-intra-node-proxy-ingress-reject.yaml]: I've reimplemented with 2 containers per pod to test allow/deny L7 paths.

Comment (joestringer) [reply examples/kubernetes/connectivity-check/pod-to-c-intra-node-proxy-ingress-reject.yaml]: Regarding the paths, it's `/public` and `/private` as those are the paths that the underlying server responds to.

File: examples/kubernetes/connectivity-check/echo-c.yaml (hunk 3 above)
Comment (joestringer) [reply examples/kubernetes/connectivity-check/echo-c.yaml]: @sayboras is there something we can do from the YAMLs to assist this?

My typical debugging involves basically `kubectl describe [...]` which hopefully tells me which pod + which container is failing, then from there I can try to reason about why that specific check will have failed.

File: examples/kubernetes/connectivity-check/proxy.cue [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+}
+_hostPortProxyPolicy: _egressL7Policy & {
+	// TODO: Is hostport translated at the source too? Then port 80. Otherwise port 40001.
+	//_port: 80
+	metadata: labels: component: "hostport-check"
+}
+deployment: "pod-to-c-multi-node-hostport-proxy-egress": _hostPortProxyResource
+egressCNP: "pod-to-c-multi-node-hostport-proxy-egress": _hostPortProxyPolicy
+deployment: "pod-to-c-intra-node-hostport-proxy-egress": _hostPortProxyResource
+egressCNP: "pod-to-c-intra-node-hostport-proxy-egress": _hostPortProxyPolicy
```
Comment (joestringer) [inline examples/kubernetes/connectivity-check/proxy.cue]: @jrajahalme I've added hostport checks here too.

These lines will generate the YAML to deploy pods with egress L7 policy per the declaration at the top of this file (effectively allowing requests to `/public`), with the destination `"echo-c-host-headless:40001"`, a service that is declared in `echo-servers.cue`. Both intra-node and multi-node checks are added here, with affinity derived from the name of the check here `*-{intra,multi}-node-*` (via `defaults.cue`).

**Comment by joestringer** [comment]
test-focus K8sConformance.*connectivity

**Comment by joestringer** [comment]
There's one failure occurring in the hostport proxy case, which is consistent with what I'm observing in a test cluster. I believe that's actually a Cilium bug. Looking into adding quarantine labels to allow us to add such YAMLs but not automatically deploy them.

**Comment by joestringer** [comment]
test-focus K8sConformance.*connectivity

**Comment by joestringer** [comment]
This works in manual testing + CI so probably good enough for review. I'll follow up on the smoke tests next week.

File: examples/kubernetes/connectivity-check/pod-to-c-intra-node-proxy-ingress-reject.yaml (hunk 1 above)
Comment (kkourt) [reply examples/kubernetes/connectivity-check/pod-to-c-intra-node-proxy-ingress-reject.yaml]: Thanks! FWIW, the server seems to also respond with empty messages on other paths, but `/public` and `/private` makes it unambiguous.

File: examples/kubernetes/connectivity-check/defaults.cue [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+	// readinessProbe target name
+	if ID =~ "^pod-to-a.*$" || ID =~ "^host-to-a.*$" {
+		_probeTarget: *"echo-a" | string
+	}
+	if ID =~ "^pod-to-b.*$" || ID =~ "^host-to-b.*$" {
+		_probeTarget: *"echo-b" | string
+	}
+
+	// Topology
```
Comment (kkourt) [inline examples/kubernetes/connectivity-check/defaults.cue:25]: AFAIU, the parameters below cannot be overridden, which seems in conflict with the top-level comment.
Maybe a comment here to note the exception makes sense?

File: examples/kubernetes/connectivity-check/resources.cue [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+	_allowDNS: *true | false
+
+	// Implicitly open DNS visibility if FQDN rule is specified.
+	_enableDNSVisibility: *false | true
+	for r in _rules if len(r.toFQDNs) > 0 {
+		_enableDNSVisibility: true
+	}
+
+	if !_allowDNS {
+	spec: egress: _rules
```
Comment (kkourt) [inline examples/kubernetes/connectivity-check/resources.cue]: nit: missing indentation? (edit: maybe it should be omitted altogether based on following line?)

File: examples/kubernetes/connectivity-check/services.cue [hunk 7]
```diff
@@ -0,0 +1,38 @@
+package connectivity_check
+
+deployment: [ID=_]: {
```
Comment (kkourt) [inline examples/kubernetes/connectivity-check/services.cue:3]: Any particular reason why this is not placed in `defaults.cue`?

In general, I think that minimizing the places where an entity (in this case `deployment`) is defined might be a good way to strike a good balance between making it easy to write and easy to understand/modify.

**Review by kkourt — APPROVED** [review]
Had a look at the CUE-based version, and everything looks great to me!

One suggestion would be to place some parts of the very informative commit messages in a README file in the `connectivity-check` directory.

For example:

- all of the description of how to use the CLI tools from connectivity-check: Add cue CLI tools
- the make targets description from connectivity-check: Support generating YAMLs via cue
- the implementation details from descriptions of  connectivity-check: Introduce proxy checks

File: examples/kubernetes/connectivity-check/defaults.cue (hunk 5 above)
Comment (joestringer) [reply examples/kubernetes/connectivity-check/defaults.cue:25]: Ah, true; they need `| string` to be overridden. Could split this section into two: One with fields that may be overwritten and one with final fields.

File: examples/kubernetes/connectivity-check/resources.cue (hunk 6 above)
Comment (joestringer) [reply examples/kubernetes/connectivity-check/resources.cue]: I think I may have only run `cue fmt` on a later commit, can fix this up.

File: examples/kubernetes/connectivity-check/services.cue (hunk 7 above)
Comment (joestringer) [reply examples/kubernetes/connectivity-check/services.cue:3]: I tried both approaches, and I didn't conclusively decide one path as superior to the other. Either way, the core definitions are in yet another file (`resources.cue`) and yet in this particular case the definitions are only relevant to this file.

But at the same time, these *will* apply globally so perhaps it's better to shift them into defaults.go. Another option is also to combine defaults.go and resources.go to put all of the code in one place (though I *think* it might be slightly clearer to have defaults.go be smaller and more concise for the type of defaults they provide).

Comment (kkourt) [reply examples/kubernetes/connectivity-check/services.cue:3]: I don't feel strongly about either case, so up to you.

How to factor definitions seems like a non-trivial aspect of cue. While the connectivity tests are fairly self-contained and I guess it does not matter much, it's a good playground to experiment and see if and how to extend cue's use.

File: examples/kubernetes/connectivity-check/echo-c.yaml (hunk 3 above)
Comment (sayboras) [reply examples/kubernetes/connectivity-check/echo-c.yaml]: Thanks for your offer, but seems unlikely we can do much.

I would love to get to bottom of this https://github.com/cilium/cilium/issues/12279, but it's not failing for my self hosted github action. Current logs we have from github actions didn't help much.

Comment (joestringer) [reply examples/kubernetes/connectivity-check/echo-c.yaml]: I've added a new commit on top that hopes to improve the debugging, and also covers dumping logs for all of the services (by iterating the cue definitions rather than listing them one-by-one).

File: examples/kubernetes/connectivity-check/services.cue (hunk 7 above)
Comment (joestringer) [reply examples/kubernetes/connectivity-check/services.cue:3]: I've added a new commit near the end to document the layout and how to interpret the files.

**Comment by joestringer** [comment]
I thought I'd found the underlying reason for the failure in smoke test (all tests that check for failure should have probe timeouts that are longer than the timeout of the probe command), but it appears that the smoke test is still failing. I've submitted https://github.com/cilium/cilium/pull/12689 to improve debuggability, so will look to merge that first then rebase & re-run the checks.

**Comment by sayboras** [comment]
> I thought I'd found the underlying reason for the failure in smoke test (all tests that check for failure should have probe timeouts that are longer than the timeout of the probe command), but it appears that the smoke test is still failing. I've submitted #12689 to improve debuggability, so will look to merge that first then rebase & re-run the checks.

Is the failure related to the change from livenessProbe to readinessProbe? I copied your additional test cases (including quarantine catagory as well) into another PR #12671, and need to pass `--set global.hostServices.enabled=true` flag. 

<details>
<summary>List of developments</summary>

```
deployment.apps/echo-a condition met
deployment.apps/echo-b condition met
deployment.apps/echo-b-host condition met
deployment.apps/echo-c condition met
deployment.apps/host-to-b-multi-node-clusterip condition met
deployment.apps/host-to-b-multi-node-headless condition met
deployment.apps/pod-to-a condition met
deployment.apps/pod-to-a-allowed-cnp condition met
deployment.apps/pod-to-a-intra-node-proxy-egress-allow condition met
deployment.apps/pod-to-a-intra-node-proxy-egress-reject condition met
deployment.apps/pod-to-a-l3-denied-cnp condition met
deployment.apps/pod-to-a-multi-node-proxy-egress-allow condition met
deployment.apps/pod-to-a-multi-node-proxy-egress-reject condition met
deployment.apps/pod-to-b-intra-node condition met
deployment.apps/pod-to-b-intra-node-nodeport condition met
deployment.apps/pod-to-b-multi-node-clusterip condition met
deployment.apps/pod-to-b-multi-node-headless condition met
deployment.apps/pod-to-b-multi-node-nodeport condition met
deployment.apps/pod-to-c-intra-node-proxy-ingress-allow condition met
deployment.apps/pod-to-c-intra-node-proxy-ingress-reject condition met
deployment.apps/pod-to-c-intra-node-proxy-to-proxy-allow condition met
deployment.apps/pod-to-c-intra-node-proxy-to-proxy-reject condition met
deployment.apps/pod-to-c-multi-node-proxy-ingress-allow condition met
deployment.apps/pod-to-c-multi-node-proxy-ingress-reject condition met
deployment.apps/pod-to-c-multi-node-proxy-to-proxy-allow condition met
deployment.apps/pod-to-c-multi-node-proxy-to-proxy-reject condition met
```

</details>

**Comment by joestringer** [comment]
> Is the failure related to the change from livenessProbe to readinessProbe? I copied your additional test cases (including quarantine catagory as well) into another PR #12671, and need to pass --set global.hostServices.enabled=true flag.

Good question, it appears not related to this. I was able to reproduce locally using kind though. Still investigating.

It's odd that `hostServices` feature would be required; AFAICT `kube-proxy` should be present in the nodes and should provision the relevant service connectivity if Cilium does not.

**Comment by sayboras** [comment]
> It's odd that `hostServices` feature would be required; AFAICT `kube-proxy` should be present in the nodes and should provision the relevant service connectivity if Cilium does not.

I was checking on this point for past few days, but unable to crack it. Probably, I was having wrong assumption that it's IPv6 related issue. Keen to hear your inputs and findings on this matter.

**Comment by joestringer** [comment]
I traced the smoke-test issue down to #12699 and added a new commit on top to mitigate.

**Comment by joestringer** [comment]
test-focus K8sConformance.*connectivity

**Comment by joestringer** [comment]
Go lint CI failure will be addressed by #12696. Smoke test is now passing. Running CI version of connectivity check to ensure it doesn't break there.

**Review by errordeveloper — COMMENTED** [review]
Didn't get a chance to have an in-depth look, but generally I think that using CUE for this is a very good idea, we should consider using it more in various testing use-cases. I also found that driving CUE from Go is fairly easy, we are doing that in the GKE test cluster operator, it's also more easily unit-testable that way.

File: .github/workflows/smoke-test.yaml [hunk 8]
```diff
@@ -139,6 +133,7 @@ jobs:
              --set global.externalIPs.enabled=true \
              --set global.nodePort.enabled=true \
              --set global.hostPort.enabled=true \
+             --set config.bpfMasquerade=false \
```
Comment (sayboras) [inline .github/workflows/smoke-test.yaml:135]: :+1:

File: examples/kubernetes/connectivity-check/Makefile [hunk 9]
```diff
@@ -1,35 +1,90 @@
...
+	--workdir /src/examples/kubernetes/connectivity-check \
+	--volume $(CURDIR)/../../..:/src \
+	--user "$(shell id -u):$(shell id -g)"
 
-all: $(DEFAULT_OUT) $(HOSTPORT_OUT) $(SINGLE_OUT)
+CUE_IMAGE := "docker.io/cuelang/cue:v0.2.2@sha256:2ea932c771212db140c9996c6fff1d236f3f84ae82add914374cee553b6fc60c"
+CUE := $(DOCKER_RUN) $(CUE_IMAGE)
 
-$(DEFAULT_OUT): $(DEFAULT_SRC)
+all: $(ALL_TARGETS)
```
Comment (sayboras) [inline examples/kubernetes/connectivity-check/Makefile:27]: nit: might need to add step to run this target and check `git diff` before running smoke test. Can be done as follow up.

File: examples/kubernetes/connectivity-check/connectivity-check.yaml [hunk 10]
```diff
@@ -1,112 +1,147 @@
...
-  ports:
-  - port: 80
-  selector:
-    name: echo-a
 ---
-apiVersion: apps/v1
-kind: Deployment
 metadata:
   name: echo-a
+  labels:
```
Comment (sayboras) [inline examples/kubernetes/connectivity-check/connectivity-check.yaml:4]: nit: we might need to update docs related to connectivity check as there are some name changes, and scope changes.

Can be done later.

File: examples/kubernetes/connectivity-check/Makefile [hunk 11]
```diff
@@ -1,35 +1,90 @@
...
+ALL_TARGETS := $(DEFAULT_OUT) $(HOSTPORT_OUT) $(SINGLE_OUT) $(PROXY_OUT) $(QUARANTINE_OUT)
 
+DOCKER_RUN := $(CONTAINER_ENGINE) container run --rm \
+	--workdir /src/examples/kubernetes/connectivity-check \
+	--volume $(CURDIR)/../../..:/src \
+	--user "$(shell id -u):$(shell id -g)"
 
-all: $(DEFAULT_OUT) $(HOSTPORT_OUT) $(SINGLE_OUT)
+CUE_IMAGE := "docker.io/cuelang/cue:v0.2.2@sha256:2ea932c771212db140c9996c6fff1d236f3f84ae82add914374cee553b6fc60c"
+CUE := $(DOCKER_RUN) $(CUE_IMAGE)
```
Comment (sayboras) [inline examples/kubernetes/connectivity-check/Makefile:25]: nit: $(QUIET) can be added to reduce verbosity, e.g. make help

File: .github/workflows/smoke-test.yaml (hunk 8 above)
Comment (sayboras) [reply .github/workflows/smoke-test.yaml:135]: nit: this one should be added in https://docs.cilium.io/en/latest/gettingstarted/kind as well.

File: examples/kubernetes/connectivity-check/Makefile [hunk 12]
```diff
@@ -1,35 +1,90 @@
...
+	cat $(SERVERS_OUT) >> $@
+
+clean:
+	@rm -f $(ALL_TARGETS)
+	@rm -f *.new
+	@rm -f *.diff
+	@rm -f *.yaml
+
+deploy:
+	kubectl apply -f connectivity-check-hostport.yaml
```
Comment (sayboras) [inline examples/kubernetes/connectivity-check/Makefile]: should it be common `connectivity-check.yaml` ?

**Review by sayboras — APPROVED** [review]
I don't know cuelang, so review generated files only.

LGTM :100:

File: examples/kubernetes/connectivity-check/defaults.cue [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+	}
+	if ID =~ "^pod-to-c.*$" || ID =~ "^host-to-c.*$" {
+		_probeTarget: *"echo-c" | string
+	}
+}
+
+// Default parameters for echo clients (final).
+deployment: [ID=_]: {
+	// Topology
+	if ID =~ "^[-_a-zA-Z0-9]*intra-node[-_a-zA-Z0-9]*$" {
```
Comment (nebril) [inline examples/kubernetes/connectivity-check/defaults.cue]: question - any particular reason why you are limiting the character set? If it's for render-time validation I think that's fine, but maybe we should validate these yamls via some kind of local validator as a build step?

**Review by nebril — APPROVED** [review]
Looks great! 

Linter failed for some reason in `pkg/k8s/apis/cilium.io/v2/types.go:41`, not sure if it's connected to this PR.

File: examples/kubernetes/connectivity-check/Makefile (hunk 12 above)
Comment (joestringer) [reply examples/kubernetes/connectivity-check/Makefile]: Good question. Here I opted for the larger YAML to try to ensure that any developers who use this target will get more information (even checks that may potentially fail depending on cluster configuration), just to reduce the likelihood that we overlook the extra checks in the hostport YAML.

File: examples/kubernetes/connectivity-check/defaults.cue (hunk 13 above)
Comment (joestringer) [reply examples/kubernetes/connectivity-check/defaults.cue]: Maybe it can just be `.*` instead, I don't have a specific motivation to restrict the characterset.

Is it possible to validate locally? I have been using the standard kubectl validation (which requires the resources such as CNP to already be configured in the connected cluster with schema validation included).

File: examples/kubernetes/connectivity-check/Makefile (hunk 11 above)
Comment (joestringer) [reply examples/kubernetes/connectivity-check/Makefile:25]: FWIW I'd need to add $(QUIET) to the individual make lines rather than in the definition here, just so that it won't get messed up if we call cue within a loop.

File: .github/workflows/smoke-test.yaml (hunk 8 above)
Comment (joestringer) [reply .github/workflows/smoke-test.yaml:135]: I've noted that down on issue #12699 to follow up.

File: examples/kubernetes/connectivity-check/connectivity-check.yaml (hunk 10 above)
Comment (joestringer) [reply examples/kubernetes/connectivity-check/connectivity-check.yaml:4]: New commit incoming for this.

File: examples/kubernetes/connectivity-check/Makefile (hunk 9 above)
Comment (joestringer) [reply examples/kubernetes/connectivity-check/Makefile:27]: Filed https://github.com/cilium/cilium/issues/12714.

**Comment by joestringer** [comment]
test-focus K8sConformance.*connectivity

**Comment by joestringer** [comment]
test-focus K8sConformance.*connectivity

File: examples/kubernetes/connectivity-check/README.md [hunk 14]
```diff
@@ -2,3 +2,35 @@
...
+  * Extra checks for various paths involving Layer 7 policy.
+
+## Developer documentation
+
+These checks are written in [CUE](https://cuelang.org/) to define various
+checks in a concise manner. The definitions for the checks are split across
+multiple files per the following logic:
+
+* `resources.cue`: The main definitions for templating all Kubernetes resources,
+  including Deployment, Service and CiliumNetworkPolicy.
```
Comment (nathanjsweet) [inline examples/kubernetes/connectivity-check/README.md]: Oxford (or serial) comma is usually deemed standard. change to: `Deployment, Service and, CiliumNetworkPolicy`

File: examples/kubernetes/connectivity-check/README.md [hunk 15]
```diff
@@ -2,3 +2,35 @@
...
+* `echo-servers.cue`: Data definitions for all `echo-*` servers used for other
+  connectivity checks.
+* `defaults.cue`: Default parameters used to define how specific checks connect
+  to particular echo servers, including selecting the probe destination;
```
Comment (nathanjsweet) [inline examples/kubernetes/connectivity-check/README.md]: I'm finding this semi-colon confusing. Could it be replace with an "and"?

**Review by nathanjsweet — CHANGES_REQUESTED** [review]
Two nits. Otherwise, it looks good.

File: examples/kubernetes/connectivity-check/README.md (hunk 14 above)
Comment (joestringer) [reply examples/kubernetes/connectivity-check/README.md]: Is the comma after `resources` supposed to be there?

File: examples/kubernetes/connectivity-check/README.md (hunk 15 above)
Comment (joestringer) [reply examples/kubernetes/connectivity-check/README.md]: it's poorly placed :-) "x and y and z" is an unusual way to list things though. I think this is another job for the Oxford comma.

**Comment by joestringer** [comment]
test-focus K8sConformance.*connectivity

**Review by nathanjsweet — APPROVED** [review]

**Comment by joestringer** [comment]
All code owner signoffs, and all CI checks that rely on connectivity checks have passed:

https://github.com/cilium/cilium/pull/12599/checks?check_run_id=925168850
https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated-Focus/355/

Marking `ready-to-merge`.

**Comment by aanm** [comment]
@joestringer I've changed the release not from `minor` to `misc` since it's not user faced.
